File: intellichem2mqtt/mqtt/discovery.py
"""Home Assistant MQTT Discovery configuration."""

import logging
from typing import Any

from ..config import MQTTConfig
from .client import MQTTClient

logger = logging.getLogger(__name__)


class DiscoveryManager:
    """Manager for Home Assistant MQTT Discovery.

    Generates and publishes discovery configs for all IntelliChem
    entities so they appear automatically in Home Assistant.
    """

    def __init__(self, mqtt_client: MQTTClient, config: MQTTConfig):
        """Initialize the discovery manager.

        Args:
            mqtt_client: Connected MQTT client
            config: MQTT configuration
        """
        self.client = mqtt_client
        self.config = config
        self._device_info = {
            "identifiers": ["intellichem_144"],
            "name": "IntelliChem",
            "manufacturer": "Pentair",
            "model": "IntelliChem",
            "suggested_area": "Pool",
        }

    def _discovery_topic(self, component: str, entity_id: str) -> str:
        """Build a discovery topic.

        Args:
            component: HA component type (sensor, binary_sensor, etc.)
            entity_id: Unique entity identifier

        Returns:
            Discovery topic string
        """
        return f"{self.config.discovery_prefix}/{component}/intellichem/{entity_id}/config"

    def _state_topic(self, *path: str) -> str:
        """Build a state topic.

        Args:
            path: Topic path components

        Returns:
            State topic string
        """
        return f"{self.config.topic_prefix}/intellichem/{'/'.join(path)}"

    def _base_config(self, name: str, entity_id: str) -> dict[str, Any]:
        """Build base discovery config.

        Args:
            name: Entity display name
            entity_id: Unique entity identifier

        Returns:
            Base config dictionary
        """
        return {
            "name": name,
            "unique_id": f"intellichem_144_{entity_id}",
            "availability_topic": self.client.availability_topic,
            "payload_available": "online",
            "payload_not_available": "offline",
            "device": self._device_info,
        }

    async def publish_discovery_configs(self) -> None:
        """Publish all discovery configs to Home Assistant."""
        logger.info("Publishing Home Assistant discovery configs")

        # Sensors
        await self._publish_sensors()

        # Binary sensors
        await self._publish_binary_sensors()

        # Text sensors for status displays
        await self._publish_text_sensors()

        logger.info("Discovery configs published")

    async def _publish_sensors(self) -> None:
        """Publish sensor discovery configs."""
        sensors = [
            # pH sensors
            {
                "name": "pH Level",
                "entity_id": "ph_level",
                "state_topic": self._state_topic("ph", "level"),
                "unit_of_measurement": "pH",
                "state_class": "measurement",
                "icon": "mdi:ph",
            },
            {
                "name": "pH Setpoint",
                "entity_id": "ph_setpoint",
                "state_topic": self._state_topic("ph", "setpoint"),
                "unit_of_measurement": "pH",
                "icon": "mdi:target",
            },
            {
                "name": "pH Tank Level",
                "entity_id": "ph_tank_level",
                "state_topic": self._state_topic("ph", "tank_level_percent"),
                "unit_of_measurement": "%",
                "icon": "mdi:car-coolant-level",
            },
            {
                "name": "pH Dose Time",
                "entity_id": "ph_dose_time",
                "state_topic": self._state_topic("ph", "dose_time"),
                "unit_of_measurement": "s",
                "device_class": "duration",
                "icon": "mdi:timer",
            },
            {
                "name": "pH Dose Volume",
                "entity_id": "ph_dose_volume",
                "state_topic": self._state_topic("ph", "dose_volume"),
                "unit_of_measurement": "mL",
                "icon": "mdi:beaker",
            },
            # ORP sensors
            {
                "name": "ORP Level",
                "entity_id": "orp_level",
                "state_topic": self._state_topic("orp", "level"),
                "unit_of_measurement": "mV",
                "device_class": "voltage",
                "state_class": "measurement",
                "icon": "mdi:flash",
            },
            {
                "name": "ORP Setpoint",
                "entity_id": "orp_setpoint",
                "state_topic": self._state_topic("orp", "setpoint"),
                "unit_of_measurement": "mV",
                "device_class": "voltage",
                "icon": "mdi:target",
            },
            {
                "name": "ORP Tank Level",
                "entity_id": "orp_tank_level",
                "state_topic": self._state_topic("orp", "tank_level_percent"),
                "unit_of_measurement": "%",
                "icon": "mdi:car-coolant-level",
            },
            {
                "name": "ORP Dose Time",
                "entity_id": "orp_dose_time",
                "state_topic": self._state_topic("orp", "dose_time"),
                "unit_of_measurement": "s",
                "device_class": "duration",
                "icon": "mdi:timer",
            },
            {
                "name": "ORP Dose Volume",
                "entity_id": "orp_dose_volume",
                "state_topic": self._state_topic("orp", "dose_volume"),
                "unit_of_measurement": "mL",
                "icon": "mdi:beaker",
            },
            # Water chemistry sensors
            {
                "name": "Temperature",
                "entity_id": "temperature",
                "state_topic": self._state_topic("temperature"),
                "unit_of_measurement": "°F",
                "device_class": "temperature",
                "state_class": "measurement",
            },
            {
                "name": "Saturation Index (LSI)",
                "entity_id": "lsi",
                "state_topic": self._state_topic("lsi"),
                "state_class": "measurement",
                "icon": "mdi:water-percent",
            },
            {
                "name": "Calcium Hardness",
                "entity_id": "calcium_hardness",
                "state_topic": self._state_topic("calcium_hardness"),
                "unit_of_measurement": "ppm",
                "state_class": "measurement",
                "icon": "mdi:flask",
            },
            {
                "name": "Cyanuric Acid",
                "entity_id": "cyanuric_acid",
                "state_topic": self._state_topic("cyanuric_acid"),
                "unit_of_measurement": "ppm",
                "state_class": "measurement",
                "icon": "mdi:flask",
            },
            {
                "name": "Alkalinity",
                "entity_id": "alkalinity",
                "state_topic": self._state_topic("alkalinity"),
                "unit_of_measurement": "ppm",
                "state_class": "measurement",
                "icon": "mdi:flask",
            },
            {
                "name": "Salt Level",
                "entity_id": "salt_level",
                "state_topic": self._state_topic("salt_level"),
                "unit_of_measurement": "ppm",
                "state_class": "measurement",
                "icon": "mdi:shaker",
            },
            {
                "name": "Firmware",
                "entity_id": "firmware",
                "state_topic": self._state_topic("firmware"),
                "icon": "mdi:chip",
            },
        ]

        for sensor in sensors:
            config = self._base_config(sensor["name"], sensor["entity_id"])
            config["state_topic"] = sensor["state_topic"]

            for key in ["unit_of_measurement", "device_class", "state_class", "icon"]:
                if key in sensor:
                    config[key] = sensor[key]

            topic = self._discovery_topic("sensor", sensor["entity_id"])
            await self.client.publish_json(topic, config, retain=True)

    async def _publish_binary_sensors(self) -> None:
        """Publish binary sensor discovery configs."""
        binary_sensors = [
            {
                "name": "Flow Detected",
                "entity_id": "flow_detected",
                "state_topic": self._state_topic("flow_detected"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "running",
                "icon": "mdi:water",
            },
            {
                "name": "Flow Alarm",
                "entity_id": "flow_alarm",
                "state_topic": self._state_topic("alarms", "flow"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
            },
            {
                "name": "pH Tank Empty",
                "entity_id": "ph_tank_empty",
                "state_topic": self._state_topic("alarms", "ph_tank_empty"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
                "icon": "mdi:car-coolant-level",
            },
            {
                "name": "ORP Tank Empty",
                "entity_id": "orp_tank_empty",
                "state_topic": self._state_topic("alarms", "orp_tank_empty"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
                "icon": "mdi:car-coolant-level",
            },
            {
                "name": "Probe Fault",
                "entity_id": "probe_fault",
                "state_topic": self._state_topic("alarms", "probe_fault"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
            },
            {
                "name": "Communication Lost",
                "entity_id": "comms_lost",
                "state_topic": self._state_topic("comms_lost"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "connectivity",
            },
            {
                "name": "pH Lockout",
                "entity_id": "ph_lockout",
                "state_topic": self._state_topic("warnings", "ph_lockout"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
            },
            {
                "name": "pH Daily Limit",
                "entity_id": "ph_daily_limit",
                "state_topic": self._state_topic("warnings", "ph_daily_limit"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
            },
            {
                "name": "ORP Daily Limit",
                "entity_id": "orp_daily_limit",
                "state_topic": self._state_topic("warnings", "orp_daily_limit"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "problem",
            },
            {
                "name": "pH Dosing",
                "entity_id": "ph_dosing",
                "state_topic": self._state_topic("ph", "is_dosing"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "running",
                "icon": "mdi:water-pump",
            },
            {
                "name": "ORP Dosing",
                "entity_id": "orp_dosing",
                "state_topic": self._state_topic("orp", "is_dosing"),
                "payload_on": "true",
                "payload_off": "false",
                "device_class": "running",
                "icon": "mdi:water-pump",
            },
        ]

        for sensor in binary_sensors:
            config = self._base_config(sensor["name"], sensor["entity_id"])
            config["state_topic"] = sensor["state_topic"]
            config["payload_on"] = sensor["payload_on"]
            config["payload_off"] = sensor["payload_off"]

            for key in ["device_class", "icon"]:
                if key in sensor:
                    config[key] = sensor[key]

            topic = self._discovery_topic("binary_sensor", sensor["entity_id"])
            await self.client.publish_json(topic, config, retain=True)

    async def _publish_text_sensors(self) -> None:
        """Publish text sensor discovery configs for status displays."""
        text_sensors = [
            {
                "name": "pH Dosing Status",
                "entity_id": "ph_dosing_status",
                "state_topic": self._state_topic("ph", "dosing_status"),
                "icon": "mdi:information",
            },
            {
                "name": "ORP Dosing Status",
                "entity_id": "orp_dosing_status",
                "state_topic": self._state_topic("orp", "dosing_status"),
                "icon": "mdi:information",
            },
            {
                "name": "Water Chemistry",
                "entity_id": "water_chemistry",
                "state_topic": self._state_topic("warnings", "water_chemistry"),
                "icon": "mdi:water-alert",
            },
        ]

        for sensor in text_sensors:
            config = self._base_config(sensor["name"], sensor["entity_id"])
            config["state_topic"] = sensor["state_topic"]
            if "icon" in sensor:
                config["icon"] = sensor["icon"]

            topic = self._discovery_topic("sensor", sensor["entity_id"])
            await self.client.publish_json(topic, config, retain=True)

    async def remove_discovery_configs(self) -> None:
        """Remove all discovery configs from Home Assistant."""
        logger.info("Removing Home Assistant discovery configs")

        entity_ids = [
            # Sensors
            "ph_level", "ph_setpoint", "ph_tank_level", "ph_dose_time", "ph_dose_volume",
            "orp_level", "orp_setpoint", "orp_tank_level", "orp_dose_time", "orp_dose_volume",
            "temperature", "lsi", "calcium_hardness", "cyanuric_acid", "alkalinity",
            "salt_level", "firmware",
            "ph_dosing_status", "orp_dosing_status", "water_chemistry",
        ]

        binary_ids = [
            "flow_detected", "flow_alarm", "ph_tank_empty", "orp_tank_empty",
            "probe_fault", "comms_lost", "ph_lockout", "ph_daily_limit",
            "orp_daily_limit", "ph_dosing", "orp_dosing",
        ]

        for entity_id in entity_ids:
            topic = self._discovery_topic("sensor", entity_id)
            await self.client.publish(topic, "", retain=True)

        for entity_id in binary_ids:
            topic = self._discovery_topic("binary_sensor", entity_id)
            await self.client.publish(topic, "", retain=True)

        logger.info("Discovery configs removed")

File: esp32/intellichem_esp32/main/mqtt/discovery.c
/**
 * @file discovery.c
 * @brief Home Assistant MQTT Discovery implementation
 */

#include "discovery.h"
#include "mqtt_task.h"
#include "publisher.h"
#include "../protocol/constants.h"
#include "esp_log.h"

#include <stdio.h>
#include <string.h>

static const char *TAG = "discovery";

// ============================================================================
// Configuration Constants
// ============================================================================

// Setpoint limits (from protocol/constants.h)
#ifndef PH_SETPOINT_MIN
#define PH_SETPOINT_MIN     7.0f
#endif
#ifndef PH_SETPOINT_MAX
#define PH_SETPOINT_MAX     7.6f
#endif
#ifndef ORP_SETPOINT_MIN
#define ORP_SETPOINT_MIN    400
#endif
#ifndef ORP_SETPOINT_MAX
#define ORP_SETPOINT_MAX    800
#endif
#ifndef CALCIUM_HARDNESS_MIN
#define CALCIUM_HARDNESS_MIN    25
#endif
#ifndef CALCIUM_HARDNESS_MAX
#define CALCIUM_HARDNESS_MAX    800
#endif
#ifndef CYANURIC_ACID_MIN
#define CYANURIC_ACID_MIN   0
#endif
#ifndef CYANURIC_ACID_MAX
#define CYANURIC_ACID_MAX   210
#endif
#ifndef ALKALINITY_MIN
#define ALKALINITY_MIN      25
#endif
#ifndef ALKALINITY_MAX
#define ALKALINITY_MAX      800
#endif

// ============================================================================
// Topic Helpers
// ============================================================================

size_t discovery_build_topic(char *buf, size_t buf_size,
                              const char *component, const char *entity_id)
{
    return snprintf(buf, buf_size, "%s/%s/intellichem/%s/config",
                    CONFIG_MQTT_DISCOVERY_PREFIX, component, entity_id);
}

size_t discovery_get_device_info(char *buf, size_t buf_size)
{
    return snprintf(buf, buf_size,
        "\"device\":{"
        "\"identifiers\":[\"intellichem_%d\"],"
        "\"name\":\"IntelliChem\","
        "\"manufacturer\":\"Pentair\","
        "\"model\":\"IntelliChem\","
        "\"suggested_area\":\"Pool\""
        "}",
        CONFIG_INTELLICHEM_ADDRESS);
}

// ============================================================================
// Internal Helper Functions
// ============================================================================

/**
 * @brief Publish a single discovery config
 */
static esp_err_t publish_discovery(esp_mqtt_client_handle_t client,
                                   const char *topic, const char *payload)
{
    int msg_id = esp_mqtt_client_publish(client, topic, payload, 0, CONFIG_MQTT_QOS, 1);
    if (msg_id < 0) {
        ESP_LOGE(TAG, "Failed to publish discovery to %s", topic);
        return ESP_FAIL;
    }
    ESP_LOGD(TAG, "Published discovery: %s", topic);
    return ESP_OK;
}

/**
 * @brief Build base config common to all entities
 */
static int build_base_config(char *buf, size_t buf_size,
                              const char *name, const char *entity_id)
{
    char device_info[256];
    discovery_get_device_info(device_info, sizeof(device_info));

    char availability_topic[128];
    publisher_get_availability_topic(availability_topic, sizeof(availability_topic));

    return snprintf(buf, buf_size,
        "\"name\":\"%s\","
        "\"unique_id\":\"intellichem_%d_%s\","
        "\"availability_topic\":\"%s\","
        "\"payload_available\":\"online\","
        "\"payload_not_available\":\"offline\","
        "%s",
        name,
        CONFIG_INTELLICHEM_ADDRESS,
        entity_id,
        availability_topic,
        device_info);
}

// ============================================================================
// Sensor Discovery
// ============================================================================

typedef struct {
    const char *name;
    const char *entity_id;
    const char *topic_path;
    const char *unit;
    const char *device_class;
    const char *state_class;
    const char *icon;
} sensor_config_t;

static const sensor_config_t SENSORS[] = {
    // pH sensors
    {"pH Level", "ph_level", "ph/level", "pH", NULL, "measurement", "mdi:ph"},
    {"pH Setpoint", "ph_setpoint", "ph/setpoint", "pH", NULL, NULL, "mdi:target"},
    {"pH Tank Level", "ph_tank_level", "ph/tank_level_percent", "%", NULL, NULL, "mdi:car-coolant-level"},
    {"pH Dose Time", "ph_dose_time", "ph/dose_time", "s", "duration", NULL, "mdi:timer"},
    {"pH Dose Volume", "ph_dose_volume", "ph/dose_volume", "mL", NULL, NULL, "mdi:beaker"},

    // ORP sensors
    {"ORP Level", "orp_level", "orp/level", "mV", "voltage", "measurement", "mdi:flash"},
    {"ORP Setpoint", "orp_setpoint", "orp/setpoint", "mV", "voltage", NULL, "mdi:target"},
    {"ORP Tank Level", "orp_tank_level", "orp/tank_level_percent", "%", NULL, NULL, "mdi:car-coolant-level"},
    {"ORP Dose Time", "orp_dose_time", "orp/dose_time", "s", "duration", NULL, "mdi:timer"},
    {"ORP Dose Volume", "orp_dose_volume", "orp/dose_volume", "mL", NULL, NULL, "mdi:beaker"},

    // Water chemistry
    {"Temperature", "temperature", "temperature", "°F", "temperature", "measurement", NULL},
    {"Saturation Index (LSI)", "lsi", "lsi", NULL, NULL, "measurement", "mdi:water-percent"},
    {"Calcium Hardness", "calcium_hardness", "calcium_hardness", "ppm", NULL, "measurement", "mdi:flask"},
    {"Cyanuric Acid", "cyanuric_acid", "cyanuric_acid", "ppm", NULL, "measurement", "mdi:flask"},
    {"Alkalinity", "alkalinity", "alkalinity", "ppm", NULL, "measurement", "mdi:flask"},
    {"Salt Level", "salt_level", "salt_level", "ppm", NULL, "measurement", "mdi:shaker"},
    {"Firmware", "firmware", "firmware", NULL, NULL, NULL, "mdi:chip"},
};

static const sensor_config_t TEXT_SENSORS[] = {
    {"pH Dosing Status", "ph_dosing_status", "ph/dosing_status", NULL, NULL, NULL, "mdi:information"},
    {"ORP Dosing Status", "orp_dosing_status", "orp/dosing_status", NULL, NULL, NULL, "mdi:information"},
    {"Water Chemistry", "water_chemistry", "warnings/water_chemistry", NULL, NULL, NULL, "mdi:water-alert"},
};

esp_err_t discovery_publish_sensors(esp_mqtt_client_handle_t client)
{
    char topic[256];
    char payload[768];
    char state_topic[128];
    char base_config[512];
    esp_err_t ret;

    ESP_LOGI(TAG, "Publishing %d sensor discovery configs",
             (int)(sizeof(SENSORS) / sizeof(SENSORS[0]) + sizeof(TEXT_SENSORS) / sizeof(TEXT_SENSORS[0])));

    // Regular sensors
    for (size_t i = 0; i < sizeof(SENSORS) / sizeof(SENSORS[0]); i++) {
        const sensor_config_t *s = &SENSORS[i];

        discovery_build_topic(topic, sizeof(topic), "sensor", s->entity_id);
        publisher_build_topic(state_topic, sizeof(state_topic), s->topic_path);
        build_base_config(base_config, sizeof(base_config), s->name, s->entity_id);

        int len = snprintf(payload, sizeof(payload),
            "{%s,\"state_topic\":\"%s\"",
            base_config, state_topic);

        if (s->unit) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"unit_of_measurement\":\"%s\"", s->unit);
        }
        if (s->device_class) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"device_class\":\"%s\"", s->device_class);
        }
        if (s->state_class) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"state_class\":\"%s\"", s->state_class);
        }
        if (s->icon) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"icon\":\"%s\"", s->icon);
        }

        snprintf(payload + len, sizeof(payload) - len, "}");

        ret = publish_discovery(client, topic, payload);
        if (ret != ESP_OK) return ret;
    }

    // Text sensors
    for (size_t i = 0; i < sizeof(TEXT_SENSORS) / sizeof(TEXT_SENSORS[0]); i++) {
        const sensor_config_t *s = &TEXT_SENSORS[i];

        discovery_build_topic(topic, sizeof(topic), "sensor", s->entity_id);
        publisher_build_topic(state_topic, sizeof(state_topic), s->topic_path);
        build_base_config(base_config, sizeof(base_config), s->name, s->entity_id);

        int len = snprintf(payload, sizeof(payload),
            "{%s,\"state_topic\":\"%s\"",
            base_config, state_topic);

        if (s->icon) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"icon\":\"%s\"", s->icon);
        }

        snprintf(payload + len, sizeof(payload) - len, "}");

        ret = publish_discovery(client, topic, payload);
        if (ret != ESP_OK) return ret;
    }

    return ESP_OK;
}

// ============================================================================
// Binary Sensor Discovery
// ============================================================================

typedef struct {
    const char *name;
    const char *entity_id;
    const char *topic_path;
    const char *device_class;
    const char *icon;
} binary_sensor_config_t;

static const binary_sensor_config_t BINARY_SENSORS[] = {
    {"Flow Detected", "flow_detected", "flow_detected", "running", "mdi:water"},
    {"Flow Alarm", "flow_alarm", "alarms/flow", "problem", NULL},
    {"pH Tank Empty", "ph_tank_empty", "alarms/ph_tank_empty", "problem", "mdi:car-coolant-level"},
    {"ORP Tank Empty", "orp_tank_empty", "alarms/orp_tank_empty", "problem", "mdi:car-coolant-level"},
    {"Probe Fault", "probe_fault", "alarms/probe_fault", "problem", NULL},
    {"Communication Lost", "comms_lost", "comms_lost", "connectivity", NULL},
    {"pH Lockout", "ph_lockout", "warnings/ph_lockout", "problem", NULL},
    {"pH Daily Limit", "ph_daily_limit", "warnings/ph_daily_limit", "problem", NULL},
    {"ORP Daily Limit", "orp_daily_limit", "warnings/orp_daily_limit", "problem", NULL},
    {"pH Dosing", "ph_dosing", "ph/is_dosing", "running", "mdi:water-pump"},
    {"ORP Dosing", "orp_dosing", "orp/is_dosing", "running", "mdi:water-pump"},
};

esp_err_t discovery_publish_binary_sensors(esp_mqtt_client_handle_t client)
{
    char topic[256];
    char payload[768];
    char state_topic[128];
    char base_config[512];
    esp_err_t ret;

    ESP_LOGI(TAG, "Publishing %d binary sensor discovery configs",
             (int)(sizeof(BINARY_SENSORS) / sizeof(BINARY_SENSORS[0])));

    for (size_t i = 0; i < sizeof(BINARY_SENSORS) / sizeof(BINARY_SENSORS[0]); i++) {
        const binary_sensor_config_t *s = &BINARY_SENSORS[i];

        discovery_build_topic(topic, sizeof(topic), "binary_sensor", s->entity_id);
        publisher_build_topic(state_topic, sizeof(state_topic), s->topic_path);
        build_base_config(base_config, sizeof(base_config), s->name, s->entity_id);

        int len = snprintf(payload, sizeof(payload),
            "{%s,\"state_topic\":\"%s\","
            "\"payload_on\":\"true\",\"payload_off\":\"false\"",
            base_config, state_topic);

        if (s->device_class) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"device_class\":\"%s\"", s->device_class);
        }
        if (s->icon) {
            len += snprintf(payload + len, sizeof(payload) - len,
                ",\"icon\":\"%s\"", s->icon);
        }

        snprintf(payload + len, sizeof(payload) - len, "}");

        ret = publish_discovery(client, topic, payload);
        if (ret != ESP_OK) return ret;
    }

    return ESP_OK;
}

// ============================================================================
// Number Entity Discovery (Control)
// ============================================================================

typedef struct {
    const char *name;
    const char *entity_id;
    const char *state_path;
    const char *command_name;
    float min;
    float max;
    float step;
    const char *unit;
    const char *icon;
    const char *mode;  // "slider" or "box"
} number_config_t;

static const number_config_t NUMBERS[] = {
    {"pH Setpoint Control", "ph_setpoint_control", "ph/setpoint", "ph_setpoint",
     PH_SETPOINT_MIN, PH_SETPOINT_MAX, 0.1f, "pH", "mdi:target", "slider"},
    {"ORP Setpoint Control", "orp_setpoint_control", "orp/setpoint", "orp_setpoint",
     ORP_SETPOINT_MIN, ORP_SETPOINT_MAX, 10.0f, "mV", "mdi:target", "slider"},
    {"Calcium Hardness Setting", "calcium_hardness_control", "calcium_hardness", "calcium_hardness",
     CALCIUM_HARDNESS_MIN, CALCIUM_HARDNESS_MAX, 25.0f, "ppm", "mdi:flask", "box"},
    {"Cyanuric Acid Setting", "cyanuric_acid_control", "cyanuric_acid", "cyanuric_acid",
     CYANURIC_ACID_MIN, CYANURIC_ACID_MAX, 10.0f, "ppm", "mdi:flask", "box"},
    {"Alkalinity Setting", "alkalinity_control", "alkalinity", "alkalinity",
     ALKALINITY_MIN, ALKALINITY_MAX, 10.0f, "ppm", "mdi:flask", "box"},
};

esp_err_t discovery_publish_number_entities(esp_mqtt_client_handle_t client)
{
    char topic[256];
    char payload[1024];
    char state_topic[128];
    char command_topic[128];
    char base_config[512];
    esp_err_t ret;

    ESP_LOGI(TAG, "Publishing %d number entity discovery configs",
             (int)(sizeof(NUMBERS) / sizeof(NUMBERS[0])));

    for (size_t i = 0; i < sizeof(NUMBERS) / sizeof(NUMBERS[0]); i++) {
        const number_config_t *n = &NUMBERS[i];

        discovery_build_topic(topic, sizeof(topic), "number", n->entity_id);
        publisher_build_topic(state_topic, sizeof(state_topic), n->state_path);
        publisher_build_command_topic(command_topic, sizeof(command_topic), n->command_name);
        build_base_config(base_config, sizeof(base_config), n->name, n->entity_id);

        snprintf(payload, sizeof(payload),
            "{%s,"
            "\"state_topic\":\"%s\","
            "\"command_topic\":\"%s\","
            "\"min\":%.1f,\"max\":%.1f,\"step\":%.1f,"
            "\"unit_of_measurement\":\"%s\","
            "\"icon\":\"%s\","
            "\"mode\":\"%s\""
            "}",
            base_config,
            state_topic, command_topic,
            n->min, n->max, n->step,
            n->unit, n->icon, n->mode);

        ret = publish_discovery(client, topic, payload);
        if (ret != ESP_OK) return ret;
    }

    return ESP_OK;
}

// ============================================================================
// Switch Entity Discovery (Control)
// ============================================================================

typedef struct {
    const char *name;
    const char *entity_id;
    const char *state_path;
    const char *command_name;
    const char *icon;
} switch_config_t;

static const switch_config_t SWITCHES[] = {
    {"pH Dosing Enable", "ph_dosing_enable", "ph/dosing_enabled", "ph_dosing", "mdi:flask-outline"},
    {"ORP Dosing Enable", "orp_dosing_enable", "orp/dosing_enabled", "orp_dosing", "mdi:flask-outline"},
};

esp_err_t discovery_publish_switch_entities(esp_mqtt_client_handle_t client)
{
    char topic[256];
    char payload[1024];
    char state_topic[128];
    char command_topic[128];
    char base_config[512];
    esp_err_t ret;

    ESP_LOGI(TAG, "Publishing %d switch entity discovery configs",
             (int)(sizeof(SWITCHES) / sizeof(SWITCHES[0])));

    for (size_t i = 0; i < sizeof(SWITCHES) / sizeof(SWITCHES[0]); i++) {
        const switch_config_t *s = &SWITCHES[i];

        discovery_build_topic(topic, sizeof(topic), "switch", s->entity_id);
        publisher_build_topic(state_topic, sizeof(state_topic), s->state_path);
        publisher_build_command_topic(command_topic, sizeof(command_topic), s->command_name);
        build_base_config(base_config, sizeof(base_config), s->name, s->entity_id);

        snprintf(payload, sizeof(payload),
            "{%s,"
            "\"state_topic\":\"%s\","
            "\"command_topic\":\"%s\","
            "\"payload_on\":\"ON\",\"payload_off\":\"OFF\","
            "\"state_on\":\"true\",\"state_off\":\"false\","
            "\"icon\":\"%s\""
            "}",
            base_config,
            state_topic, command_topic,
            s->icon);

        ret = publish_discovery(client, topic, payload);
        if (ret != ESP_OK) return ret;
    }

    return ESP_OK;
}

// ============================================================================
// Main Discovery Functions
// ============================================================================

esp_err_t discovery_publish_all(esp_mqtt_client_handle_t client, bool control_enabled)
{
    ESP_LOGI(TAG, "Publishing Home Assistant discovery configs (control=%s)",
             control_enabled ? "enabled" : "disabled");

    esp_err_t ret;

    // Sensors
    ret = discovery_publish_sensors(client);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to publish sensor discovery");
        return ret;
    }

    // Binary sensors
    ret = discovery_publish_binary_sensors(client);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to publish binary sensor discovery");
        return ret;
    }

    // Control entities (if enabled)
    if (control_enabled) {
        ret = discovery_publish_number_entities(client);
        if (ret != ESP_OK) {
            ESP_LOGE(TAG, "Failed to publish number entity discovery");
            return ret;
        }

        ret = discovery_publish_switch_entities(client);
        if (ret != ESP_OK) {
            ESP_LOGE(TAG, "Failed to publish switch entity discovery");
            return ret;
        }

        ESP_LOGI(TAG, "Control entities published");
    }

    ESP_LOGI(TAG, "All discovery configs published successfully");
    return ESP_OK;
}

esp_err_t discovery_remove_all(esp_mqtt_client_handle_t client)
{
    ESP_LOGI(TAG, "Removing all discovery configs");

    char topic[256];

    // Remove sensors
    for (size_t i = 0; i < sizeof(SENSORS) / sizeof(SENSORS[0]); i++) {
        discovery_build_topic(topic, sizeof(topic), "sensor", SENSORS[i].entity_id);
        esp_mqtt_client_publish(client, topic, "", 0, CONFIG_MQTT_QOS, 1);
    }
    for (size_t i = 0; i < sizeof(TEXT_SENSORS) / sizeof(TEXT_SENSORS[0]); i++) {
        discovery_build_topic(topic, sizeof(topic), "sensor", TEXT_SENSORS[i].entity_id);
        esp_mqtt_client_publish(client, topic, "", 0, CONFIG_MQTT_QOS, 1);
    }

    // Remove binary sensors
    for (size_t i = 0; i < sizeof(BINARY_SENSORS) / sizeof(BINARY_SENSORS[0]); i++) {
        discovery_build_topic(topic, sizeof(topic), "binary_sensor", BINARY_SENSORS[i].entity_id);
        esp_mqtt_client_publish(client, topic, "", 0, CONFIG_MQTT_QOS, 1);
    }

    // Remove number entities
    for (size_t i = 0; i < sizeof(NUMBERS) / sizeof(NUMBERS[0]); i++) {
        discovery_build_topic(topic, sizeof(topic), "number", NUMBERS[i].entity_id);
        esp_mqtt_client_publish(client, topic, "", 0, CONFIG_MQTT_QOS, 1);
    }

    // Remove switch entities
    for (size_t i = 0; i < sizeof(SWITCHES) / sizeof(SWITCHES[0]); i++) {
        discovery_build_topic(topic, sizeof(topic), "switch", SWITCHES[i].entity_id);
        esp_mqtt_client_publish(client, topic, "", 0, CONFIG_MQTT_QOS, 1);
    }

    ESP_LOGI(TAG, "All discovery configs removed");
    return ESP_OK;
}
